granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -78,7 +101,94 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
+        if self.has_parent_cgroup(cgroup_name):
+            new_cgroup_path = self._get_parent_cgroup(cgroup_name)
+        else:
+            new_cgroup_path = self.path / cgroup_name
+            new_cgroup_path.mkdir(exist_ok=True)
+
+        return new_cgroup_path
+
+    def has_parent_cgroup(self, cgroup_name: str) -> bool:
+        """
+        Check if cgroup has the given cgroup_name in its hierarchy
+        :param cgroup_name: the cgroup name
+        :return: bool indicating if cgroup is in hierarchy
+        """
+        return self._get_parent_cgroup(cgroup_name) is not None
+
+    def get_pids_in_cgroup(self) -> set[int]:
+        return {int(proc) for proc in self.read_from_interface_file(CGROUP_PROCS_FILE).split()}
+
+    def assign_process_to_cgroup(self, pid: int = 0) -> None:
+        self.write_to_interface_file(CGROUP_PROCS_FILE, str(pid))
+
+    def read_from_interface_file(self, interface_name: str) -> str:
+        interface_path = self.path / interface_name
+        return interface_path.read_text()
+
+    def write_to_interface_file(self, interface_name: str, data: str) -> None:
+        interface_path = self.path / interface_name
+        interface_path.write_text(data)
+
+    @abstractmethod
+    def get_cgroup_in_hierarchy(self, controller: str, cgroup_name: str) -> CgroupCore:
+        """
+        Return a CGroup in the hierarchy which has the given name.
+        If one doesn't exist higher up in the hierarchy, create a new one lower in the hierarchy.
+        """
+        pass
+
+
+class CgroupCoreV1(CgroupCore):
+    def get_cgroup_in_hierarchy(self, controller: str, cgroup_name: str) -> CgroupCore:
+        subcgroup_path = self._get_cgroup_in_hierarchy(cgroup_name)
+        assert subcgroup_path is not None
+        return CgroupCoreV1(subcgroup_path)
+
+
+CGROUP_V2_SUPPORTED_CONTROLLERS = "cgroup.controllers"
+CGROUP_V2_ENABLED_CONTROLLERS = "cgroup.subtree_control"
+
+
+class CgroupCoreV2(CgroupCore):
+    def is_controller_supported(self, controller: str):
+        return controller in self.read_from_interface_file(CGROUP_V2_SUPPORTED_CONTROLLERS).split()
+
+    def enable_controller(self, controller: str):
+        assert self.is_controller_supported(controller), f"Controller not supported {controller!r}"
+        enabled_controllers_file = self.path / CGROUP_V2_ENABLED_CONTROLLERS
+        enabled_controllers_file.write_text(f"+{controller}{os.linesep}")
+
+    def disable_controller(self, controller: str):
+        enabled_controllers_file = self.path / CGROUP_V2_ENABLED_CONTROLLERS
+        enabled_controllers_file.write_text(f"-{controller}{os.linesep}")
+
+    def get_cgroup_in_hierarchy(self, controller: str, cgroup_name: str) -> CgroupCore:
+        subcgroup_path = self._get_cgroup_in_hierarchy(cgroup_name)
+        assert subcgroup_path is not None
+
+        new_cgroup = CgroupCoreV2(subcgroup_path)
+        new_cgroup.enable_controller(controller)
+        return new_cgroup
```

Comment:
So this, logically, will either enable a controller for our current `cgroup_path`, or will enable it for a `cgroup_path` that was created under us.

Is this comment relevant?
![image](https://user-images.githubusercontent.com/8831572/215291400-3e0db7cd-bddc-4e6d-bd37-0ab7b6f5878f.png)

i.e, if the CURRENT cgroup don't have the controller enabled, then a child cgroup won't be able to enable it, to my understanding.

Reply:
Discussed F2F, the solution we decided on was trying to create the new cGroup under the parent of current cGroup and if we cannot (e.g. the controller isn't delegated to the parent), we try to create it under the mount point

Reply:
Resolving as this is no longer relevant